services/codeclimate/codeclimate-analysis.service.js

```diff
@@ -0,0 +1,178 @@
+'use strict'
+
+const Joi = require('joi')
+const { BaseJsonService, NotFound } = require('..')
+const { colorScale, letterScore } = require('../color-formatters')
+const { nonNegativeInteger } = require('../validators')
+const { keywords, isLetterGrade, fetchRepo } = require('./codeclimate-common')
+
+const schema = Joi.object({
+  data: Joi.object({
+    meta: Joi.object({
+      issues_count: nonNegativeInteger,
+    }).required(),
+    attributes: Joi.object({
+      ratings: Joi.array()
+        .items(
+          Joi.object({
+            letter: isLetterGrade,
+            measure: Joi.object({
+              value: Joi.number().required(),
+            }).required(),
+          })
+        )
+        .length(1)
+        .required(),
+    }).required(),
+  }).required(),
+}).required()
+
+const maintainabilityColorScale = colorScale(
+  [50, 80, 90, 95],
+  ['red', 'yellow', 'yellowgreen', 'green', 'brightgreen']
+)
+const techDebtColorScale = colorScale(
+  [5, 10, 20, 50],
+  ['brightgreen', 'green', 'yellowgreen', 'yellow', 'red']
+)
+const issueColorScale = colorScale(
+  [1, 5, 10, 20],
+  ['brightgreen', 'green', 'yellowgreen', 'yellow', 'red']
+)
+
+const whichMap = {
+  maintainability: {
+    transform: data => ({
+      maintainabilityLetter: data.attributes.ratings[0].letter,
+    }),
+    render: ({ maintainabilityLetter }) => ({
+      label: 'maintainability',
+      message: maintainabilityLetter,
+      color: letterScore(maintainabilityLetter),
+    }),
+  },
+  'maintainability-percentage': {
+    transform: data => ({
+      techDebtPercentage: data.attributes.ratings[0].measure.value,
+    }),
+    render: ({ techDebtPercentage }) => {
+      // maintainability = 100 - technical debt.
+      const maintainabilityPercentage = 100 - techDebtPercentage
+      return {
+        label: 'maintainability',
+        message: `${maintainabilityPercentage.toFixed(0)}%`,
+        color: maintainabilityColorScale(maintainabilityPercentage),
+      }
+    },
+  },
+  'tech-debt': {
+    transform: data => ({
+      techDebtPercentage: data.attributes.ratings[0].measure.value,
+    }),
+    render: ({ techDebtPercentage }) => ({
+      label: 'technical debt',
+      message: `${techDebtPercentage.toFixed(0)}%`,
+      color: techDebtColorScale(techDebtPercentage),
+    }),
+  },
+  issues: {
+    transform: data => ({
+      issueCount: data.meta.issues_count,
+    }),
+    render: ({ issueCount }) => ({
+      label: 'issues',
+      message: `${issueCount}`,
+      color: issueColorScale(issueCount),
+    }),
+  },
+}
+
+module.exports = class CodeclimateAnalysis extends BaseJsonService {
+  static get route() {
+    return {
+      base: 'codeclimate',
+      pattern:
+        ':which(maintainability|maintainability-percentage|tech-debt|issues)/:user/:repo',
```

Comment:
Do we know what the relationship is between these 3 metrics? Based on the above calculation it seems maintainability and tech-debt are directly related, but what about issues? 

Thinking along the lines of https://github.com/badges/shields/issues/3174, I sometimes find myself wondering when/how to group certain metrics together (note I'm not suggesting this class be split, this class just made me remember that question/thought)  

I could potentially see down the line someone wanting to display different forms of issues (for example based on [severity](https://docs.codeclimate.com/docs/issues#section-issue-severity)) ![](https://img.shields.io/badge/issues-1%20blocker%2C%202%20minor-red.svg). In that hypothetical, what would the considerations be for adding that inline vs. creating a new class? Would we only split the class if there was a need for a new route parameter?

Reply:
I've never used Code Climate, though I think issues is showing the number of issues with the code, and the tech debt and maintainability scores are derived from the issue count. So they seem closely related. I was thinking that post-#3174 it would be sufficient to have a single entry in Analysis.

I think the alternate formatting for issues could be done with a query parameter, though if different `which`es support different query parameters, we may need to consider that as part of #3174. Alternatively it may make more sense to make them separate services.

Reply:
Thanks that helps!